src/transformStateWithClones.js

```diff
@@ -7,7 +7,52 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const history = [];
+  let previousState = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        previousState = createStateWithAddedProperties(
+          previousState,
+          action.extraData,
+        );
+        break;
+      case 'removeProperties':
+        previousState = createStateWithRemovedProperties(
+          previousState,
+          action.keysToRemove,
+        );
+        break;
+      case 'clear':
+        previousState = createClearState(previousState);
+        break;
+      default:
+        throw new Error(`Unknown action type: ${action.type}`);
+    }
+
+    history.push({ ...previousState });
+  }
+
+  return history;
```

Comment:
The `transformStateWithClones` function is correctly not mutating the original `state` object and is using a switch statement with a default case for error handling, which is good practice.